File: gateway/src/routes/coherence.ts
/**
 * Coherence Routes - Proxy to Coherence Service
 *
 * Provides endpoints for:
 * - Project coherence status
 * - Conflict management
 * - Rule information
 */

import express, { Response, Router } from 'express';
import axios, { AxiosError } from 'axios';
import { authenticate, requireAuth, AuthRequest } from '../middleware/auth.js';
import { logger } from '../utils/logger.js';

const router: Router = express.Router();

// Coherence service URL from environment
const COHERENCE_SERVICE_URL = process.env.COHERENCE_SERVICE_URL || 'http://coherence-service:8000';
const INTERNAL_API_KEY = process.env.INTERNAL_API_KEY || '';

// Axios client for coherence service
const coherenceClient = axios.create({
  baseURL: COHERENCE_SERVICE_URL,
  timeout: 30000,
});

/**
 * Build headers for coherence service requests.
 */
function buildHeaders(req: AuthRequest): Record<string, string> {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'X-Internal-API-Key': INTERNAL_API_KEY,
  };

  if (req.user) {
    headers['X-User-ID'] = req.user.id;
    headers['X-User-Email'] = req.user.email;
    headers['X-User-Role'] = req.user.role;
    if (req.user.institutionId) {
      headers['X-Institution-ID'] = req.user.institutionId;
    }
  }

  return headers;
}

/**
 * Handle errors from coherence service.
 */
function handleError(error: unknown, res: Response, context: string): void {
  if (axios.isAxiosError(error)) {
    const axiosError = error as AxiosError;
    const status = axiosError.response?.status || 500;
    const data = axiosError.response?.data || { error: 'Coherence service error' };

    logger.error(`Coherence ${context} error:`, {
      status,
      data,
      message: axiosError.message,
    });

    res.status(status).json(data);
  } else {
    logger.error(`Coherence ${context} error:`, error);
    res.status(500).json({ error: 'Internal server error' });
  }
}

// ============================================================================
// Status Endpoints
// ============================================================================

/**
 * GET /api/coherence/projects/:projectId/status
 * Get coherence status for a project.
 */
router.get(
  '/projects/:projectId/status',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { projectId } = req.params;

    try {
      const response = await coherenceClient.get(
        `/api/coherence/projects/${projectId}/status`,
        { headers: buildHeaders(req) }
      );
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'get status');
    }
  }
);

/**
 * POST /api/coherence/projects/:projectId/validate
 * Trigger full validation of a project.
 */
router.post(
  '/projects/:projectId/validate',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { projectId } = req.params;

    try {
      const response = await coherenceClient.post(
        `/api/coherence/projects/${projectId}/validate`,
        req.body,
        { headers: buildHeaders(req) }
      );
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'validate');
    }
  }
);

/**
 * POST /api/coherence/projects/:projectId/check-realtime
 * Real-time coherence check for a fact change.
 */
router.post(
  '/projects/:projectId/check-realtime',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { projectId } = req.params;

    try {
      const response = await coherenceClient.post(
        `/api/coherence/projects/${projectId}/check-realtime`,
        req.body,
        {
          headers: buildHeaders(req),
          timeout: 5000, // Quick timeout for hot-path
        }
      );
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'realtime check');
    }
  }
);

// ============================================================================
// Conflict Endpoints
// ============================================================================

/**
 * GET /api/coherence/projects/:projectId/conflicts
 * Get all conflicts for a project.
 */
router.get(
  '/projects/:projectId/conflicts',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { projectId } = req.params;
    const { status } = req.query;

    try {
      const params = status ? { status } : {};
      const response = await coherenceClient.get(
        `/api/coherence/projects/${projectId}/conflicts`,
        {
          headers: buildHeaders(req),
          params,
        }
      );
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'get conflicts');
    }
  }
);

/**
 * GET /api/coherence/projects/:projectId/conflicts/:conflictId
 * Get a specific conflict.
 */
router.get(
  '/projects/:projectId/conflicts/:conflictId',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { projectId, conflictId } = req.params;

    try {
      const response = await coherenceClient.get(
        `/api/coherence/projects/${projectId}/conflicts/${conflictId}`,
        { headers: buildHeaders(req) }
      );
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'get conflict');
    }
  }
);

/**
 * POST /api/coherence/projects/:projectId/conflicts/:conflictId/resolve
 * Resolve a conflict.
 */
router.post(
  '/projects/:projectId/conflicts/:conflictId/resolve',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { projectId, conflictId } = req.params;

    try {
      const response = await coherenceClient.post(
        `/api/coherence/projects/${projectId}/conflicts/${conflictId}/resolve`,
        req.body,
        { headers: buildHeaders(req) }
      );
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'resolve conflict');
    }
  }
);

// ============================================================================
// Rule Endpoints
// ============================================================================

/**
 * GET /api/coherence/rules
 * List all coherence rules.
 */
router.get(
  '/rules',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { enabled_only, category } = req.query;

    try {
      const response = await coherenceClient.get('/api/coherence/rules', {
        headers: buildHeaders(req),
        params: { enabled_only, category },
      });
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'list rules');
    }
  }
);

/**
 * GET /api/coherence/rules/:ruleId
 * Get a specific rule.
 */
router.get(
  '/rules/:ruleId',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { ruleId } = req.params;

    try {
      const response = await coherenceClient.get(
        `/api/coherence/rules/${ruleId}`,
        { headers: buildHeaders(req) }
      );
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'get rule');
    }
  }
);

/**
 * POST /api/coherence/rules/:ruleId/evaluate
 * Evaluate a single rule against a project.
 */
router.post(
  '/rules/:ruleId/evaluate',
  authenticate,
  requireAuth,
  async (req: AuthRequest, res: Response) => {
    const { ruleId } = req.params;
    const { project_id } = req.query;

    try {
      const response = await coherenceClient.post(
        `/api/coherence/rules/${ruleId}/evaluate`,
        req.body,
        {
          headers: buildHeaders(req),
          params: { project_id },
        }
      );
      res.json(response.data);
    } catch (error) {
      handleError(error, res, 'evaluate rule');
    }
  }
);

export default router;

File: coherence-service/app/routers/coherence.py
"""Coherence API Router.

Provides endpoints for:
- Project coherence status
- Conflict listing and resolution
- Rule management
- Real-time and batch validation
"""

import logging
from typing import Any, Dict, List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Header
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_settings
from app.models.rules import CoherenceRule, RuleEvaluation
from app.models.conflicts import (
    Conflict,
    ConflictResolution,
    ConflictStatus,
    CoherenceStatus,
)
from app.services.rule_engine import get_rule_engine
from app.services.conflict_detector import get_conflict_detector

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/coherence", tags=["coherence"])

# Settings
settings = get_settings()


# ============================================================================
# Dependencies
# ============================================================================

async def get_db():
    """Get database session - placeholder for dependency injection."""
    # This will be properly configured in main.py
    from app.database import get_async_session
    async for session in get_async_session():
        yield session


async def verify_internal_key(
    x_internal_api_key: Optional[str] = Header(None),
) -> bool:
    """Verify internal API key for service-to-service calls."""
    if not settings.INTERNAL_API_KEY:
        return True
    if x_internal_api_key != settings.INTERNAL_API_KEY:
        raise HTTPException(status_code=401, detail="Invalid internal API key")
    return True


# ============================================================================
# Request/Response Models
# ============================================================================

class RealtimeCheckRequest(BaseModel):
    """Request for real-time coherence check."""
    fact_key: str = Field(..., description="The fact key being changed")
    new_value: Any = Field(..., description="The new value")
    source: str = Field(..., description="Source of the change")


class RealtimeCheckResponse(BaseModel):
    """Response from real-time check."""
    conflicts: List[Conflict]
    checked_in_ms: float


class ValidationRequest(BaseModel):
    """Request for full validation."""
    persist_conflicts: bool = Field(default=True, description="Whether to save detected conflicts")


class RulesResponse(BaseModel):
    """List of coherence rules."""
    rules: List[CoherenceRule]
    total: int


# ============================================================================
# Status Endpoints
# ============================================================================

@router.get("/projects/{project_id}/status", response_model=CoherenceStatus)
async def get_project_coherence_status(
    project_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: bool = Depends(verify_internal_key),
):
    """Get overall coherence status for a project.

    Returns:
        CoherenceStatus with score, conflict counts, and top issues
    """
    detector = get_conflict_detector(db)
    return await detector.get_coherence_status(project_id)


@router.post("/projects/{project_id}/validate", response_model=CoherenceStatus)
async def validate_project(
    project_id: UUID,
    request: ValidationRequest = ValidationRequest(),
    db: AsyncSession = Depends(get_db),
    _: bool = Depends(verify_internal_key),
):
    """Trigger full validation of a project.

    Runs all coherence rules and returns updated status.
    """
    detector = get_conflict_detector(db)
    return await detector.check_full(project_id, persist=request.persist_conflicts)


@router.post("/projects/{project_id}/check-realtime", response_model=RealtimeCheckResponse)
async def check_realtime(
    project_id: UUID,
    request: RealtimeCheckRequest,
    db: AsyncSession = Depends(get_db),
    _: bool = Depends(verify_internal_key),
):
    """Perform real-time coherence check for a fact change.

    This is the hot-path endpoint that must complete quickly (<100ms).
    Used when a user changes a value to immediately detect conflicts.
    """
    import time
    start = time.time()

    detector = get_conflict_detector(db)
    conflicts = await detector.check_realtime(
        project_id=project_id,
        fact_key=request.fact_key,
        new_value=request.new_value,
        source=request.source,
    )

    elapsed = (time.time() - start) * 1000
    return RealtimeCheckResponse(conflicts=conflicts, checked_in_ms=elapsed)


# ============================================================================
# Conflict Endpoints
# ============================================================================

@router.get("/projects/{project_id}/conflicts", response_model=List[Conflict])
async def get_project_conflicts(
    project_id: UUID,
    status: Optional[ConflictStatus] = Query(None, description="Filter by status"),
    db: AsyncSession = Depends(get_db),
    _: bool = Depends(verify_internal_key),
):
    """Get all conflicts for a project.

    Args:
        status: Optional filter by conflict status (active, resolved, etc.)
    """
    detector = get_conflict_detector(db)
    return await detector.get_conflicts(project_id, status=status)


@router.get("/projects/{project_id}/conflicts/{conflict_id}", response_model=Conflict)
async def get_conflict(
    project_id: UUID,
    conflict_id: str,
    db: AsyncSession = Depends(get_db),
    _: bool = Depends(verify_internal_key),
):
    """Get details of a specific conflict."""
    detector = get_conflict_detector(db)
    conflicts = await detector.get_conflicts(project_id)

    for conflict in conflicts:
        if conflict.conflict_id == conflict_id:
            return conflict

    raise HTTPException(status_code=404, detail="Conflict not found")


@router.post("/projects/{project_id}/conflicts/{conflict_id}/resolve", response_model=Conflict)
async def resolve_conflict(
    project_id: UUID,
    conflict_id: str,
    resolution: ConflictResolution,
    x_user_id: Optional[str] = Header(None),
    db: AsyncSession = Depends(get_db),
    _: bool = Depends(verify_internal_key),
):
    """Resolve a conflict.

    Accepts a resolution choice and optionally propagates changes.
    """
    user_id = x_user_id or "system"

    detector = get_conflict_detector(db)
    result = await detector.resolve_conflict(
        conflict_id=conflict_id,
        resolution=resolution,
        user_id=user_id,
    )

    if not result:
        raise HTTPException(status_code=404, detail="Conflict not found")

    return result


# ============================================================================
# Rule Endpoints
# ============================================================================

@router.get("/rules", response_model=RulesResponse)
async def list_rules(
    enabled_only: bool = Query(True, description="Only return enabled rules"),
    category: Optional[str] = Query(None, description="Filter by category"),
    _: bool = Depends(verify_internal_key),
):
    """List all coherence rules.

    Returns the rules that the coherence engine checks against.
    """
    engine = get_rule_engine()

    if category:
        rules = engine.get_rules_by_category(category)
    else:
        rules = engine.get_all_rules(enabled_only=enabled_only)

    return RulesResponse(rules=rules, total=len(rules))


@router.get("/rules/{rule_id}", response_model=CoherenceRule)
async def get_rule(
    rule_id: str,
    _: bool = Depends(verify_internal_key),
):
    """Get a specific rule by ID."""
    engine = get_rule_engine()
    rule = engine.get_rule(rule_id)

    if not rule:
        raise HTTPException(status_code=404, detail="Rule not found")

    return rule


@router.post("/rules/{rule_id}/evaluate", response_model=RuleEvaluation)
async def evaluate_rule(
    rule_id: str,
    project_id: UUID = Query(..., description="Project to evaluate against"),
    db: AsyncSession = Depends(get_db),
    _: bool = Depends(verify_internal_key),
):
    """Evaluate a single rule against a project.

    Useful for testing rules or checking specific conditions.
    """
    engine = get_rule_engine()
    rule = engine.get_rule(rule_id)

    if not rule:
        raise HTTPException(status_code=404, detail="Rule not found")

    detector = get_conflict_detector(db)
    kb = await detector._get_knowledge_base(project_id)

    if not kb:
        raise HTTPException(status_code=404, detail="Project knowledge base not found")

    result = engine.evaluate_rule(rule, kb)
    return result
